_posts/2022-1-19-quantization-in-practice.md

```diff
@@ -0,0 +1,199 @@
+torch.quantization.prepare(m, inplace=True)
+
+"""Calibrate"""
+with torch.inference_mode():
+  for _ in range(10):
+    x = torch.rand(1,2,28)
+    m(x)
+    
+"""Convert"""
+torch.quantization.convert(m, inplace=True)
+
+"""Check"""
+print(m[1].weight().element_size()) # 1 instead of 4 for FP32
+
+
+## FX GRAPH
+from torch.quantization import quantize_fx
+qconfig_dict = {"": torch.quantization.get_default_qconfig('fbgemm')}
+model_to_quantize.eval()
+# Prepare
+model_prepared = quantize_fx.prepare_fx(model_to_quantize, qconfig_dict)
+# Calibrate 
+with torch.inference_mode():
+  for _ in range(10):
+    x = torch.rand(1,2,28)
+    model_prepared(x)
+# quantize
+model_quantized = quantize_fx.convert_fx(model_prepared)
+```
+
+### Quantization-aware Training (QAT)
+The previous two methods are to quantize FP32 models *after* they have been trained. Although they work surprisingly well ([citation to some case study](google.com)), they're still subject to prediction errors arising from the drop in numerical precision.
+
+<p align="center">
+<img src="/assets/images/ptq_vs_qat.png" alt="Fig. 6: Comparison of PTQ and QAT" width="100%">
+<br>
+Wu, Hao, et al. "Integer quantization for deep learning inference: Principles and empirical evaluation." arXiv preprint arXiv:2004.09602 (2020)
+</p>
+
+
+Figure 6(a) shows why. In PTQ, the FP32 model's parameters are chosen by optimizing on the FP32 loss, and then projected to INT8 space. Depending on where the new INT8 weights lie on the loss curve, model accuracies can significantly change.
+
+In QAT, the FP32 parameters are chosen by also optimizing on the INT8 loss. This allows the model to identify a wider region in the loss function (Figure 6(b)), and identify FP32 parameters such that quantizing them does not significantly affect accuracy.
+
+It's likely that you can still use QAT by "fine-tuning" it on a sample of the training dataset, but I did not try it on demucs (yet).
+
+
+## Quantizing "real-world" models
+
+**Download the [notebook](https://gist.github.com/suraj813/735357e56321237950a0348b50f2f3b4) or run it on [Colab](https://colab.research.google.com/gist/suraj813/735357e56321237950a0348b50f2f3b4/fx-and-eager-mode-quantization-example.ipynb) (note that Colab runtimes may differ significantly from local machines).**
+
+Traceable models can be easily quantized with FX Graph Mode, but it's possible the model you're using is not traceable end-to-end. Maybe it has loops or `if` statements on inputs (dynamic control flow), or relies on third-party libraries. The model I use in this example has [dynamic control flow and uses third-party libraries](https://github.com/facebookresearch/demucs/blob/v2/demucs/model.py). As a result, it cannot be symbolically traced directly. In this code walkthrough, I show how you can bypass this limitation by quantizing the child modules individually for FX Graph Mode, and how to patch Quant/DeQuant stubs in Eager Mode.
+
+
+
+## What's next - Define-by-Run Quantization
```

Comment:
This would be in the followup post too and you can basically gett people excited about reading this followup post by describing briefly how cool it is and how it works

Also make sure to synthesize everything people learnt at the very end, that your table and figures will help this doc be a reference for people new to quantization